react/features/speaker-stats/middleware.js

```diff
@@ -0,0 +1,62 @@
+// @flow
+
+import {
+    PARTICIPANT_JOINED,
+    PARTICIPANT_KICKED,
+    PARTICIPANT_LEFT,
+    PARTICIPANT_UPDATED
+} from '../base/participants/actionTypes';
+import { MiddlewareRegistry } from '../base/redux';
+
+import { INIT_SEARCH, INIT_UPDATE_STATS } from './actionTypes';
+import { reorderStats, updateStats } from './actions';
+import { SPEAKER_STATS_RELOAD_INTERVAL } from './constants';
+import { filterBySearchCriteria, getSortedSpeakerStats } from './functions';
+
+let reorderTimeoutHandle;
+
+MiddlewareRegistry.register(({ dispatch, getState }) => next => action => {
+    const result = next(action);
+
+    switch (action.type) {
+
+    case INIT_SEARCH: {
+        const state = getState();
+        const stats = filterBySearchCriteria(state);
+
+        dispatch(updateStats(stats));
+        break;
+    }
+
+    case INIT_UPDATE_STATS:
+    case PARTICIPANT_JOINED:
+    case PARTICIPANT_LEFT:
+    case PARTICIPANT_KICKED:
+    case PARTICIPANT_UPDATED: {
+        let reorderFlag = true;
+
+        if (action.type === INIT_UPDATE_STATS && action.getSpeakerStats) {
+            const state = getState();
+            const speakerStats = { ...action.getSpeakerStats() };
+            const stats = filterBySearchCriteria(state, speakerStats);
+
+            dispatch(updateStats(stats));
+
+            reorderFlag = Boolean(action.reorder);
+        }
+
+        if (reorderFlag) {
+            clearTimeout(reorderTimeoutHandle);
+            reorderTimeoutHandle = setTimeout(() => {
+                const newState = getState();
+
+                dispatch(reorderStats(getSortedSpeakerStats(newState)));
```

Comment:
When executing this, do you expect the previous call of dispatch ^ (`dispatch(updateStats(stats));`) to had finished ... I'm not sure this is the case here.

Reply:
I believe the previous call will be finished after the 1s delay, but still I think I found a better solution that does not depend on a timeout and the reordering is initiated with the next periodic update.